File: examples/python/examples/discover_gloves.py
from __future__ import annotations

from qnbot_sdk import SerialConnection
from qnbot_sdk.glove import GloveConfig, discover_gloves


def main() -> None:
    discovered = discover_gloves()
    for device in discovered:
        if device.error is not None:
            print(
                f"port={device.port} error={device.error.code.value} "
                f"message={device.error.message}"
            )
            continue
        print(
            f"port={device.port} hand={device.hand.value} "
            f"sn={device.sn} activated={device.activated}"
        )

    selected = next((device for device in discovered if device.error is None), None)
    if selected is None or selected.hand is None:
        return

    config = GloveConfig(
        side=selected.hand,
        connection=SerialConnection(port=selected.port),
    )
    print(f"selected port={config.connection.port} hand={config.side.value}")


if __name__ == "__main__":
    main()

File: examples/cpp/examples/discover_gloves.cpp
#include <qnbot/glove.hpp>

#include <algorithm>
#include <cstdlib>
#include <iostream>

int main() {
    try {
        const auto devices = qnbot::discover_gloves();
        for (const auto& device : devices) {
            if (device.error) {
                std::cout << "port=" << device.port
                          << " error=" << device.error->message << '\n';
                continue;
            }
            std::cout << "port=" << device.port << " sn=" << *device.sn
                      << " activated=" << std::boolalpha << *device.activated
                      << '\n';
        }

        const auto selected =
            std::find_if(devices.begin(), devices.end(),
                         [](const auto& item) { return !item.error; });
        if (selected == devices.end() || !selected->hand) return EXIT_SUCCESS;

        qnbot::SerialConnection connection;
        connection.port = selected->port;
        const qnbot::GloveConfig config{*selected->hand, connection};
        std::cout << "selected port=" << selected->port << " hand="
                  << (config.side == qnbot::Side::left ? "left" : "right")
                  << '\n';
        return EXIT_SUCCESS;
    } catch (const std::exception& error) {
        std::cerr << error.what() << '\n';
        return EXIT_FAILURE;
    }
}
